src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexIgnoreCaseTests.cs

```diff
@@ -0,0 +1,221 @@
+            Regex regex = await RegexHelpers.GetRegexAsync(engine, pattern, RegexOptions.IgnoreCase);
+            Assert.True(regex.IsMatch($"{firstChar}{secondChar}"));
+            Assert.True(regex.IsMatch($"{secondChar}{firstChar}"));
+        }
+
+        [Theory]
+        [MemberData(nameof(EnginesThatSupportBackreferences))]
+        public async Task Ensure_CultureInvariant_Option_Is_Used_For_Backreferences(RegexEngine engine)
+        {
+            using var _ = new ThreadCultureChange("tr-TR");
+            Regex regex = await RegexHelpers.GetRegexAsync(engine, @"(.)\1", RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);
+            // There is no mapping between 'i' and 'I' in tr-TR culture, so this test is validating that when passing CultureInvariant
+            // option, we will use InvariantCulture mappings for backreferences.
+            Assert.True(regex.IsMatch("iI"));
+        }
+
+        [Fact]
+        // This test creates a source generated engine for each of the ~870 cultures and ensures the result compiles. This test alone takes around 30
+        // seconds on a fast machine, so marking as OuterLoop.
+        [OuterLoop]
+        public async Task SourceGenerator_Supports_All_Cultures()
+        {
+            foreach (CultureInfo culture in CultureInfo.GetCultures(CultureTypes.AllCultures))
+            {
+                using (_ = new ThreadCultureChange(culture))
+                {
+                    // This test will try to emit code that looks like: textInfo = CultureInfo.GetCultureInfo(CurrentCulture.Name).TextInfo
+                    // so we will validate in this test that we are able to do that for all cultures and that GetCultureInfo returns a valid Culture.
+                    _ = await RegexHelpers.GetRegexAsync(RegexEngine.SourceGenerated, @"(.)\1", RegexOptions.IgnoreCase);
+                }
+            }
+        }
+
+        // This test takes a long time to run since it needs to compute all possible lowercase mappings across
+        // 3 different cultures and then creates Regex matches for all of our engines for each mapping.
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(Unicode_IgnoreCase_TestData))]
+        public async Task Unicode_IgnoreCase_Tests(RegexEngine engine, string culture, RegexOptions options)
+        {
+            var testCases = GetPatternAndInputsForCulture(culture);
+
+            foreach ((string pattern, string input) in testCases)
+            {
+
+                if ((options & RegexOptions.CultureInvariant) == 0)
+                {
+                    using var _ = new ThreadCultureChange(culture);
+                    await ValidateMatch(pattern, input);
+                    await ValidateMatch(input, pattern);
+                }
+                else
+                {
+                    await ValidateMatch(pattern, input);
+                    await ValidateMatch(input, pattern);
+                }
```

Comment:
Rather than duplicating the body, you could do something like:
```suggestion
                using (ThreadCultureChange c = (options & RegexOptions.CultureInvariant) == 0 ? new ThreadCultureChange(culture) : null)
                {
                    await ValidateMatch(pattern, input);
                    await ValidateMatch(input, pattern);
                }
```